design-proposals/003-credentials.md

```diff
@@ -0,0 +1,158 @@
+Considering that OAuth requires per-user and per-application secrets, there is a need for system (Syndesis owned) and user (citizen user owned) credentials. Melding of those would be also responsibility of credential provider.
+Management of user and system credentials is also performed by credential provider, for instance initial setup of system credentials or refresh of user credentials.
+Credential provider consists of:
+ * **handler** - to implement OAuth flow (REST)
+ * **applicator** - to apply credentials to connection properties
+ * **deployer** - to augment k8s deployment descriptor
+ * **manager** - to manage credentials
+
+## Domain
+
+There are two types of domain objects, differentiated by security: public data for UI and sensitive secrets for the backend.
+
+Credentials need to be combined or linked together with other credentialsm, this has merits when considering that OAuth credentials are partialy owned by the user and partialy owned by the system. For example Syndesis owns non-public credentials (OAuth Client Secret, X.509 certificate and private key) and semi-public (OAuth Client ID), that need to be used during acquisition and in a running integration together with private credentials associated with the user (username, OAuth authorization token, OAuth refresh token). This makes for two credentials with two different security and lifecycle requirements.
+
+NOTE: Salesforce supports using x509 certificates for OAuth flow, and that should be prefered as the refreshing OAuth token can be performed within certificate's expiry i.e. not depending on the refresh token.
+
+### Credential view model
+
+Each credential should have a name so that the user can identify it, type: this can be username-password, x509 or OAuth tokens, internal identifier and connector label so it can be determined to what connection it can be applied to.
+
+The credential name can be auto generated (e.g. "Salesforce credentials for user@example.com").
+
+View model credential properties:
+
+ * **id**        : System assigned identifier for the credential, e.g. UUID or sequence
+ * **name**      : Name of the credential, for humans, e.g. "Salesforce credentials for user@example.com"
+ * **connector** : Conector type (label) that the credential applies to, e.g. "salesforce"
+ * **created**   : Date/time the credential was created
+ * **note**      : Optional note, could be generated during credential acquisition, e.g. "Expires 23.11.2018"
+ 
+This constititutes the view model, notice that the sensitive (secret) data is not exposed to the UI -- the UI should reference credentials by the identifier.
+
+### Credential backend model
+
+Credential backend model is key-value pairs grouped together under the same identifier. The identifier is identical in value to the identifier used in the view model. Keys of the pairs correspond to the property names needed for the Spring Boot configuration of the Camel components.
```

Comment:
This assumes that the connection credentials are stored separate from the connection, which I think we discussed and decided we didn't need to do?

Only system wide private credentials (oauth client id/secret) should be stored separately and not leaked to client.